proof/access-control/AARCH64/ArchArch_AC.thy

```diff
@@ -2233,6 +2233,49 @@ lemma vcpu_invalidate_active_respects[wp]:
    \<lbrace>\<lambda>_. integrity aag X st\<rbrace>"
   unfolding integrity_no_hyp_absorb by (wpsimp wp: hoare_vcg_all_lift)
 
+\<comment> \<open>Flush\<close>
+
+lemma vcpu_invalidate_active_vcpu_proj[wp]:
+  "\<lbrace>\<lambda>s. vopt = vcpu_proj r l True v a p s \<and> savedWhenDisabledRegs \<subseteq> r \<and> h\<rbrace>
+   vcpu_invalidate_active
+   \<lbrace>\<lambda>rv s. vopt = vcpu_proj r l h v a p s\<rbrace>"
+  supply if_split[split del] if_split_asm[split]
+  unfolding vcpu_invalidate_active_def
+  by wpsimp
+
+\<comment> \<open>FIXME: trivial helper lemmas for vcpu_flush_integrity_hyp\<close>
+lemma vcpu_invalid_active_arm_current_vcpu_not_Some[wp]:
```

Comment:
Do you mean with the FIXME that these lemmas might not be necessary or that they should be moved elsewhere?

They could go into AInvs, but if I think about fixing the FIXME, I'm not sure I would actually decide to move them if they are not used there currently.

Reply:
Ok, the first one has been replaced with `strengthen None_Some_strg` and the second one is now done by using an instantiated `hoare_pre_cont` directly in the relevant proof. Now that I'm thinking about it again, I'd say that the `False` precondition was a safe `wp` lemma but could have caused some serious confusion if that lemma applied unexpectedly elsewhere.

Reply:
As an aside, while adding that `strengthen` I thought that it would be nice if it could be added to `wpsimp` directly instead of needing to do `(wpsimp ... | strengthen ...)+`. Unfortunately, it looks like `strengthen` can't be used in Eisbach methods, because it doesn't seem to like the dummy invocation I think it does.

```
exception CTERM raised (line 149 of "l4v/lib/Monads/Strengthen.thy"):
  gather_to_imp
  TERM _
  ```

Reply:
Huh, interesting. I've also wondered a few times about adding strengthen to wpsimp, you got one step further with that than me :-). Maybe we can make Strengthen.thy more robust by catching the exception and defaulting to no_tac if we get one.